File: old/server/db/mongo/upsert.go
package mongo

import (
	"context"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo/options"
	"log"
)

func Upsert(collection string, data interface{}, filter bson.D) error {
	db := getDatabase()
	coll := db.Collection(collection)

	update := bson.D{{"$set", data}}
	opts := options.Update().SetUpsert(true)
	_, err := coll.UpdateOne(context.TODO(), filter, update, opts)
	if err != nil {
		log.Fatalf("Failed to upsert list: %v", err)
	}

	return nil
}

File: server/src/db/mongodb/upsert.ts
import {getDatabase} from "./instance";
import {Document, Filter, UpdateFilter} from "mongodb";


export const upsert = async <Data extends object>(collection: string, data: Data, filter: Filter<Document>): Promise<Data> => {
    const db = getDatabase();
    const mongoCollection = db.collection(collection)
    const update: UpdateFilter<unknown> = {$set: data}

    try {
        await mongoCollection.updateOne(filter, update, {upsert: true})

        return data
    } catch (e) {
        throw Error(`Failed to upsert in collection "${collection}" with data: ${JSON.stringify(data)}. ${e}`)
    }
};